sql/expression/div.go

```diff
@@ -290,20 +290,58 @@ func (d *Div) div(ctx *sql.Context, lval, rval interface{}) (interface{}, error)
 	return nil, errUnableToCast.New(lval, rval)
 }
 
-// floatOrDecimalType returns either Float64 or decimaltype depending on column reference,
-// left and right expressions types and left and right evaluated types.
+// determineResultType looks at the expressions in the expression tree with this division operation and determines
+// the result type of this division expression. This involves looking at the types of the expressions in the tree,
+// and looking for float types or Decimal types. If |outermostResult| is false, then we prefer to treat ints as floats
+// (instead of Decimals) for performance reasons, but when |outermostResult| is true, we must treat ints as Decimals
+// in order to match MySQL's behavior.
+func (d *Div) determineResultType(outermostResult bool) sql.Type {
+	//TODO: what if both BindVars? should be constant folded
+	rTyp := d.Right.Type()
+	if types.IsDeferredType(rTyp) {
+		return rTyp
+	}
+	lTyp := d.Left.Type()
+	if types.IsDeferredType(lTyp) {
+		return lTyp
+	}
+
+	if types.IsText(lTyp) || types.IsText(rTyp) {
+		return types.Float64
+	}
+
+	// For division operations, the result type is always either a float or decimal.Decimal. When working with
+	// integers, we prefer float types internally, since the performance is orders of magnitude faster to divide
+	// floats than to divide Decimals, but if this is the outermost division operation, we need to
+	// return a decimal in order to match MySQL's results exactly.
+	if outermostResult {
+		return floatOrDecimalType(d, false)
+	} else {
+		return floatOrDecimalType(d, true)
+	}
```

Comment:
This can be simplified to
```suggestion
	return floatOrDecimalType(d, !outermostResult)
```

Reply:
oh! nice catch! thank u! 😄